### PR TITLE
Autonomous routine that balances on a charge station in the end

### PR DESCRIPTION
**Pull request by DensityInfinite**
This routine will score a cone before moving to the charge station.
All poses need to be tested on the robot.

### CODE REVIEW

File: autonomous/score_and_balance.py [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+        return self
+
+
+@dataclass
+class FieldPose:
+    pose: Pose2d
+    approach_angle: Rotation2d
+    intermediate_waypoints: tuple[Translation2d, ...]
+
+    def with_correct_fipped(self) -> "FieldPose":
```
Comment (auscompgeek) [inline autonomous/score_and_balance.py:61]: what's a fipped?

File: autonomous/score_and_balance.py [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+        elif self.movement.is_at_goal():
+            self.next_state("balance")
+        self.movement.do_autodrive()
+
+    @state
+    def balance(self, initial_call: bool, tm: float) -> None:
+        if initial_call:
+            self.movement.toggle_balance()
+        elif not self.movement.is_executing:
+            print(f"Finished auto at {tm}")
```
Comment (auscompgeek) [inline autonomous/score_and_balance.py:171]: Autonomous state machines already log when they finish.
```suggestion
```

File: autonomous/score_and_balance.py [hunk 3]
```diff
@@ -0,0 +1,210 @@
...
+    gripper: Gripper
+
+    INTAKE_PRE_TIME = 2.5
+    SCORE_PRE_TIME = 2.5
+    MANUAL_CUBE_TIME = 0.5
+
+    def __init__(self) -> None:
+        self.pickup_actions: list[PickupAction] = []
+        self.score_actions: list[ScoreAction] = []
+        self.station_location: list[FieldPose] = []
```
Comment (auscompgeek) [inline autonomous/score_and_balance.py:90]: what's a station location, and why is it a list?

File: autonomous/score_and_balance.py [hunk 4]
```diff
@@ -0,0 +1,210 @@
...
+    recover: RecoverController
+    arm_component: Arm
+    gripper: Gripper
+
+    INTAKE_PRE_TIME = 2.5
+    SCORE_PRE_TIME = 2.5
+    MANUAL_CUBE_TIME = 0.5
+
+    def __init__(self) -> None:
+        self.pickup_actions: list[PickupAction] = []
```
Comment (auscompgeek) [inline autonomous/score_and_balance.py:88]: is this used?

File: autonomous/score_and_balance.py [hunk 5]
```diff
@@ -0,0 +1,210 @@
...
+        self.score_actions: list[ScoreAction] = []
```
Comment (auscompgeek) [inline autonomous/score_and_balance.py:89]: this sounds like it should _not_ have a default, and should always be passed in from the subclass, like
```py
def __init__(self) -> None:
	super().__init__(score_actions=[...])
```

**Review by auscompgeek — CHANGES_REQUESTED** [review]

File: autonomous/score_and_balance.py (hunk 1 above)
Comment (DensityInfinite) [reply autonomous/score_and_balance.py:61]: Turns out I couldn't spell, should have been "flipped"

Comment (auscompgeek) [reply autonomous/score_and_balance.py:61]: may I suggest enabling spell check in your editor?

Comment (DensityInfinite) [reply autonomous/score_and_balance.py:61]: Yea okay

**Pull request merged**